ateam_geometry/src/segment.cpp

```diff
@@ -0,0 +1,169 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "ateam_geometry/segment.hpp"
+#include "ateam_geometry/utilities.hpp"
+#include <Eigen/Dense>
+#include <math.h>
+#include <optional>
+#include <tuple>
+
+namespace ateam_geometry {
+    LineSegment::LineSegment(const Eigen::Vector2d & start, const Eigen::Vector2d & end) {
+        p1 = start;
+        p2 = end;
+
+        // Length is Euclidian, we can change later if desired
+        double length = std::sqrt(std::pow(p1.x() - p2.x(), 2) + std::pow(p1.y() - p2.y(), 2));
+    }
+    
+    Eigen::Vector2d LineSegment::get_midpoint(){
+        Eigen::Vector2d midpoint;
+        midpoint.x() = (p1.x() + p2.x()) / 2;
+        midpoint.y() = (p1.y() + p2.y()) / 2;
+        return midpoint;
+    }
+
+    std::vector<Eigen::Vector2d> LineSegment::get_equally_spaced_points(const int & num_points) {
+        /*
+        Get equally spaced points along the line segement this is called from.
+        If we only get one point, it will be the first endpoint.
+        */
+        std::vector<Eigen::Vector2d> points;
+        double x_spacing = p2.x() - p1.x() / num_points;
+        double y_spacing = p2.y() - p1.y() / num_points;
+        for (int i; i < num_points; ++i) {
+            Eigen::Vector2d point;
+            point.x() = p1.x() + (x_spacing * i);
+            point.y() = p1.y() + (y_spacing * i);
+            points.push_back(point);
+        }
+        return points;
+    }
+
+    LineSegment get_lineseg_of_length_from_point(const Eigen::Vector2d & start, const double & length, const double & angle) {
+        Eigen::Vector2d endpoint;
+        // Essentially we convert a given polar coordinate/vector 
+        // (length and angle) to an x, y plane
+        endpoint.x() =  start.x() + (length * cos(angle));
+        endpoint.y() = start.y() + (length * sin(angle));
+        
+        LineSegment segment = LineSegment(start, endpoint);
+
+        return segment;
+    }
+
+    bool is_point_on_segment(const Eigen::Vector2d & point, const LineSegment & segment) {
+        /*
+        https://computergraphics.stackexchange.com/questions/2105/test-if-a-point-is-on-a-line-segment
+        https://lucidar.me/en/mathematics/check-if-a-point-belongs-on-a-line-segment/
+        */
+        Eigen::Vector2d direction; // Projection of B onto A
+        direction.x() = segment.p2.x() - segment.p1.x();
+        direction.y() = segment.p2.y() - segment.p2.y();
+        direction.normalize();
+
+        Eigen::Vector2d p_relative; // projection of P onto A
+        p_relative.x() = point.x() - segment.p1.x();
+        p_relative.y() = point.y() - segment.p1.y();
+        double projection = direction.dot(p_relative);
+        
+        if (projection < 0 || projection > segment.length) {
+            return false;
+        }
+        return true;
```

Comment:
We also need to check that the distance between the projection point and the input point is 0 I think

Reply:
Isn't it just on the line (on the first endpoint) in that case, which would return true?

Reply:
![image](https://github.com/SSL-A-Team/software/assets/5797304/d459d207-24c0-46a4-bf2c-ae44df46e208)

I don't think that's guaranteed. Only blue in the image above should return true, but I think the current code will allow blue or green to return true. Specifically for green, we will find the projection of green onto the segment and just see if that projection is within the bounds of the segment length. Green can be thousands of meters away, but as long as it's perpendicular to the segment, it will return true I think.

See the last 6 paragraphs of the answer [here](https://computergraphics.stackexchange.com/a/2108)

Reply:
Sorry, I think I misunderstood your original comment and was thinking you were referring to changing line 88. This check makes sense

Reply:
Okay, I think this should be better now.

Reply:
Jk that was very wrong but maybe it's better now?

Reply:
This should be tested and work